### ISSUE TITLE
Need more graceful errors if ENVSETUP_ENVIRON isn't set.

### ISSUE THREAD

**Issue by mstackhouse** [pharmaverse/envsetup#10]
### What happened?

It looks like `envsetup::read_path()` will fail if you don't set ENVSETUP_ENVIRON:

Let's say I have the config file:

```yaml
default:
  paths:
    data: './DEV/username/project1/data'
    output: './demo/DEV/username/project1/output'
    programs: './demo/DEV/username/project1/programs'
```

Following the basics to setup the Rprofile, this is the error returned:

```r
envsetup::read_path("data", "test.sas7bdat")
Error in which(names(lib) == envsetup_environ):length(lib) : 
  argument of length 0
```

My impression is this comes from here:
https://github.com/pharmaverse/envsetup/blob/539514d89cbff4f61635052ba865d05a1d3fb181/R/paths.R#L28

Which is clear by the error, but in this config, we didn't name the vectors for the libraries, so we get back a 0 length vector:

``` r
my_path <- file.path('a', 'b' ,'c')
print(my_path)
#> [1] "a/b/c"

print(which(names(my_path) == ""))
#> integer(0)

which(names(my_path) == ""):1
#> Error in which(names(my_path) == ""):1: argument of length 0
```

<sup>Created on 2022-10-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So there needs to be a condition around assigning `restircted_paths` that checks that:
 - The library you're calling is actually named
 - Additionally, we might want a default assumption if envsetup_environ is not enabled. 

@nicholas-masel curious of your thoughts around this. 

### Session Information

R version 4.1.1 (2021-08-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 22000)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

loaded via a namespace (and not attached):
 [1] compiler_4.1.1   magrittr_2.0.3   assertthat_0.2.1 cli_3.4.1        envsetup_0.0.1   tools_4.1.1      rstudioapi_0.14 
 [8] yaml_2.3.6       config_0.3.1     rlang_1.0.6      renv_0.15.2      purrr_0.3.5

### Reproducible Example

``` r
config_path <- "C://Users//16105//OneDrive - ATorus//Documents//Projects//envsetup_test/_envsetup.yml"

envsetup_config <- config::get(file = config_path)

envsetup::rprofile(envsetup_config)

objects("envsetup:paths")
#> [1] "data"     "output"   "programs"

Sys.getenv("ENVSETUP_ENVIRON")
#> [1] "DEFAULT"

print(output)
#> [1] "."

envsetup::read_path(output, "my_file.txt")
#> Error in which(names(lib) == envsetup_environ):length(lib): argument of length 0
```

<sup>Created on 2022-10-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by nicholas-masel** [comment]
@mstackhouse  Thanks for the nice example.

I see [`set_autos()`](https://github.com/pharmaverse/envsetup/blob/main/R/autos.R#L35) already handles this better than `read_path()`.

```r
  if (envsetup_environ %in% names(autos_paths)) {
    autos_paths <-
      autos_paths[which(names(autos_paths) == envsetup_environ):length(autos_paths)]
  }
````

I think we can add this condition around the `read_path()` list subsetting.  So if users don't name the list and control the execution environment, it will check all paths, all the time.

This is fine for `read_path()`, but we need just one path for `write_path()`.  We can add the check to `write_path()` to throw a nice error if `lib` contains more than one path and `envsetup_environ` is not being used to name the execution evironment, i.e. telling this where to write.

Another thought would be a new helper function to run on config files to sus this out during config creation.

What do you think?  And any thoughts on documentation updates to make this more clear?

**Comment by mstackhouse** [comment]
> I think we can add this condition around the read_path() list subsetting. So if users don't name the list and control the execution environment, it will check all paths, all the time.

I think that sounds good, particularly if people are just using unnamed defaults. But I think this actually looks like an issue for the autos paths too:

``` r
envsetup_environ <- getOption("ENVSETUP_ENVIRON")
paths <- file.path("unnamed", "path")

if (envsetup_environ %in% names(paths)) {
  print("Blah")
}
#> Error in if (envsetup_environ %in% names(paths)) {: argument is of length zero
```

<sup>Created on 2022-10-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So what we can do here is:

``` r
envsetup_environ <- getOption("ENVSETUP_ENVIRON")
paths <- file.path("unnamed", "path")

if (!is.null(envsetup_environ) && envsetup_environ %in% names(paths)) {
  print("Blah")
}
```

<sup>Created on 2022-10-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by mstackhouse** [comment]
> Another thought would be a new helper function to run on config files to sus this out during config creation.

So like a config validator - I think we're homing in on a few rules here that would be valid checks so that would be really useful

**Comment by mstackhouse** [comment]
> And any thoughts on documentation updates to make this more clear?

I'll need to do a review here, but since there's only one vignette, I think we probably want to split vignettes out to outline different potential configurations. For example:
- Using envsetup for single path environments
- Using envsetup for hierarchical path environments
- A vignette around reading and writing data
- A vignette around using functions stored in autos paths

Another thing I think could be more clear is _when_ is the right time to call `library(envsetup)`. I think I created an infinite loop in an Rprofile I was working with yesterday, because it crashed my R session every time I restarted. I didn't have time to pinpoint the issue though - and this was with calling envsetup in the .RProfile itself.

**Comment by mstackhouse** [comment]
@nicholas-masel  another question - do we have a configuration option to have separate levels but _not_ look up the hierarchy? In your language, what if QA should _just_ look at QA and PROD should _just_ look at PROD?

**Comment by nicholas-masel** [comment]
> @nicholas-masel another question - do we have a configuration option to have separate levels but _not_ look up the hierarchy? In your language, what if QA should _just_ look at QA and PROD should _just_ look at PROD?

@mstackhouse  Yes, this is doable if we are ok with updating `write_path()` just write to the first path in the hierarchy.  

Then we would create the config file with default left blank but add hardcoded configurations for DEV, QA and PROD where QA doesn't have DEV paths, and PROD only has PROD paths.  Now you just select the correct configuration when calling [`config::get()`](https://rstudio.github.io/config/reference/get.html) using the config parameter.

For example:

```r
config::get(file = config_path, config = "PDEV")
config::get(file = config_path, config = "QA")
config::get(file = config_path, config = "PROD")
```

**Comment by mstackhouse** [comment]
@nicholas-masel ah ok gotcha - I probably didn't read the vignette thoroughly enough but that makes much more sense now.

**Comment by nicholas-masel** [comment]
Closed by #52
